library/bigip_user.py

```diff
@@ -182,411 +172,358 @@
             return True
-        else:
-            return False
+        self.create_on_device()
+        return True
 
-    def is_description_changed(self, current):
-        full_name = self.params['full_name']
-        if full_name is None:
+    def update(self):
+        self.have = self.read_current_from_device()
+        if not self.should_update():
             return False
-        if 'full_name' not in current:
-            return True
-        if full_name != current['full_name']:
+        if self.client.check_mode:
             return True
-        else:
-            return False
+        self.update_on_device()
+        return True
 
-    def ensure_user_is_present(self):
-        if not self.params['password_credential']:
-            raise F5ModuleError(
-                "A password_credential must be specified"
-            )
-        params = self.get_user_creation_parameters()
-        self.changed_params = camel_dict_to_snake_dict(params)
-        if self.params['check_mode']:
-            return True
-        self.create_user_on_device(params)
-        if self.user_exists():
+    def remove(self):
+        if self.client.check_mode:
             return True
-        else:
-            raise F5ModuleError("Failed to create the user")
-
-    def get_user_creation_parameters(self):
-        result = dict(
-            name=self.params['username_credential'],
-            partition=self.params['partition'],
-            partitionAccess=self.determine_partition_access_to_create()
-        )
-        if self.params['full_name']:
-            result['description'] = self.params['full_name']
-        if self.params['password_credential']:
-            result['password'] = self.params['password_credential']
-        if not self.params['shell']:
-            return result
-        if self.params['shell'] == 'none':
-            return result
-        if self.params['shell'] == 'tmsh':
-            return result
-        if self.can_have_advanced_shell_upon_creation():
-            result['shell'] = self.params['shell']
-        if self.params['shell'] == 'bash':
-            raise F5ModuleError(
-                "Custom shells are only available to administrators"
-            )
-
-    def can_have_advanced_shell_upon_creation(self):
-        roles_with_advanced_shell = [
-            'admin', 'resource-admin'
-        ]
-        if 'partitionAccess' not in self.params:
-            return False
-
-        for x in self.params['partitionAccess']:
-            if x['role'] in roles_with_advanced_shell:
-                return True
-        return False
+        self.remove_from_device()
+        if self.exists():
+            raise F5ModuleError("Failed to delete the user")
+        return True
 
-    def create_user_on_device(self, params):
-        tx = self.api.tm.transactions.transaction
-        with TransactionContextManager(tx) as api:
-            api.tm.auth.users.user.create(**params)
+    def remove_from_device(self):
```

Comment:
These _on|from_device codes should reflect the pattern that is used in GTM wide_ip.

In that module, you'll find a ModuleManager that is relatively small and which uses a factory method to create different concrete Managers based on the version of BIG-IP it detects.

This allows for two things at least

1. that we can expand this module if they break something in the future without having to worry about breaking the existing versions
2. so that we can customize future, or existing, versions with other executable code if we find that something needs to be refactored or changed.

This is an example of how the module manager would look

https://github.com/F5Networks/f5-ansible/blob/devel/library/bigip_gtm_wide_ip.py#L217

and these are examples of the two specialized classes.

https://github.com/F5Networks/f5-ansible/blob/devel/library/bigip_gtm_wide_ip.py#L336
https://github.com/F5Networks/f5-ansible/blob/devel/library/bigip_gtm_wide_ip.py#L383

Note that they inherit from a BaseManager class that includes code general to each specialized manager.